File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int result;
        int k_copy = k;
        dfs(root, k_copy, result);
        return result;
    }

    void dfs (TreeNode* root, int& k, int& result) {
        if (root->left) dfs(root->left, k, result);
        if (k == 0) return;
        k--;
        if (k == 0) {
            result = root->val;
            return;
        }
        if (root->right) dfs(root->right, k, result);
        return;
    }



};

File: Data Structures & Algorithms/valid-parenthesis-string/submission-1.py
class Solution :
    def dfs(self, i :int, open :int, s :str, memo :dict[int, dict[int, int]]) -> bool:
        if i == len(s): 
            if open == 0:
                return True
            else:
                return False
        
        if i in memo and open in memo[i]:
            return memo[i][open]
        result = False
        
        if s[i] == '(':
            result = self.dfs(i+1, open+1, s, memo)
        elif s[i] == ')':
            if open <= 0:
                result = False
            else:
                result =  self.dfs(i+1, open - 1, s, memo)
        elif s[i] == '*':
            if open >= 0:
                result = result or self.dfs(i+1, open -1, s, memo)
            result = result or self.dfs(i+1, open, s, memo)
            result = result or self.dfs(i+1, open+1, s, memo)
        if i not in memo:
            memo[i] = {}
        memo[i][open] = result
        return result

    def checkValidString(self, s :str) -> bool:
        if s == '':
            return True
        memo :dict[int, dict[int, int]] = {}
        return self.dfs(0, 0, s, memo)

      
